src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs

```diff
@@ -101,5 +104,54 @@ public AcquireTokenByAuthorizationCodeParameterBuilder WithPkceCodeVerifier(stri
             Parameters.PkceCodeVerifier = pkceCodeVerifier;
             return this;
         }
+
+        /// <summary>
+        /// To help with resiliency, AAD Cached Credential Service (CCS) operates as an AAD backup.
+        /// /// This will provide CCS with a routing hint to help improve performance during authentication.
+        /// </summary>
+        /// <param name="userObjectIdentifier">GUID which is unique to the user, parsed from the client_info.</param>
+        /// <param name="tenantIdentifier">GUID format of the tenant ID, parsed from the client_info.</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public AcquireTokenByAuthorizationCodeParameterBuilder WithCcsRoutingHint(string userObjectIdentifier, string tenantIdentifier)
+        {
+            if (string.IsNullOrEmpty(userObjectIdentifier))
+            {
+                throw new ArgumentException(MsalErrorMessage.CcsRoutingHintMissing, nameof(userObjectIdentifier));
+            }
+            if (string.IsNullOrEmpty(tenantIdentifier))
+            {
+                throw new ArgumentException(MsalErrorMessage.CcsRoutingHintMissing, nameof(tenantIdentifier));
+            }
+
+            var ccsRoutingHeader = new Dictionary<string, string>() 
+            {
+                { Constants.CcsRoutingHintHeader, CoreHelpers.GetCcsClientInfoHint(userObjectIdentifier, tenantIdentifier) } 
+            };
+
+            this.WithExtraHttpHeaders(ccsRoutingHeader);
+            return this;
+        }
+
+        /// <summary>
+        /// To help with resiliency, AAD Cached Credential Service (CCS) operates as an AAD backup.
+        /// This will provide CCS with a routing hint to help improve performance during authentication.
+        /// </summary>
+        /// <param name="userName">Identifier of the user. Generally in UserPrincipalName (UPN) format, e.g. <c>john.doe@contoso.com</c></param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public AcquireTokenByAuthorizationCodeParameterBuilder WithCcsRoutingHint(string userName)
```

Comment:
Not all parameter builders have WithLoginHint so I am adding this to here and on the OBO parameter builder